src/Umbraco.Cms.Api.Management/Mapping/Permissions/DocumentPermissionMapper.cs

```diff
@@ -56,17 +89,51 @@ public IEnumerable<IGranularPermission> MapToGranularPermissions(IPermissionPres
             };
             yield break;
         }
+
         foreach (var verb in documentPermissionPresentationModel.Verbs)
         {
             if (string.IsNullOrEmpty(verb))
             {
                 continue;
             }
+
             yield return new DocumentGranularPermission
             {
                 Key = documentPermissionPresentationModel.Document.Id,
                 Permission = verb,
             };
         }
     }
+
+    /// <inheritdoc/>
+    public IEnumerable<IPermissionPresentationModel> AggregatePresentationModels(IUser user, IEnumerable<IPermissionPresentationModel> models)
+    {
+        // Get the unique document keys that have granular permissions.
+        IEnumerable<Guid> documentKeysWithGranularPermissions = models
+            .Cast<DocumentPermissionPresentationModel>()
+            .Select(x => x.Document.Id)
+            .Distinct();
+
+        // Batch retrieve all documents by their keys.
+        var documents = _contentService.Value.GetByIds(documentKeysWithGranularPermissions)
+            .ToDictionary(doc => doc.Key, doc => doc.Path);
```

Comment:
This might be premature optimization, but maybe we should retrieve `IEntitySlim` from the entity service instead of retrieving the entire document, since all we use is the path and key, which `IEntitySlim` also contains 😄